### PR TITLE
Rate limits

### PR DESCRIPTION
**Pull request by oleurud**
Implement rate limits features for:
- https://github.com/CartoDB/Windshaft-cartodb/issues/820
- https://github.com/CartoDB/CartoDB-SQL-API/issues/459

### CODE REVIEW

File: lib/carto_metadata.js [hunk 1]
```diff
@@ -520,10 +524,123 @@ module.exports = function(redis_opts) {
...
+            me.rate_limits_db,
+            me.rate_limits_sha_script ? 'EVALSHA' : 'EVAL',
+            params,
+            function (err, rateLimits) {
+                if (err) {
+                    if (err.name === 'ReplyError' && err.message === 'NOSCRIPT No matching script. Please use EVAL.') {
+                        me.rate_limits_sha_script = null;
+                        return me.getRateLimit(username, app, endpointGroup, callback);
+                    } else {
+                        callback(err);
```
Comment (dgaubert) [inline lib/carto_metadata.js]: I'm missing a `return` here, aren't you?

File: lib/carto_metadata.js [hunk 2]
```diff
@@ -520,10 +524,123 @@ module.exports = function(redis_opts) {
...
+        );
+    };
+
+    /**
+     * Returns the inner rateLimit what is the strictest one
+     * @param {Array} rateLimits Each inner array has 5 integers indicating: 
+     *      isBloqued, limit, remaining, retry, reset
+     */
+    me.getLowerRateLimit = function(rateLimits) {
+        /*jshint maxcomplexity:10 */
```
Comment (dgaubert) [inline lib/carto_metadata.js]: Do we use `jshint` as linter in this project?

File: lib/carto_metadata.js (hunk 1 above)
Comment (dgaubert) [inline lib/carto_metadata.js]: I'm missing a `return` here. Aren't you?

File: lib/carto_metadata.js (hunk 2 above)
Comment (dgaubert) [inline lib/carto_metadata.js]: Do we use `jshint` as linter in this project?

File: lib/carto_metadata.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (oleurud) [reply lib/carto_metadata.js]: Good catch! :scream:

File: lib/carto_metadata.js [hunk 3]
```diff
@@ -520,6 +524,117 @@ module.exports = function(redis_opts) {
...
+                break;
+            }
+    
+            if (minRemainingValue === undefined || remaining < minRemainingValue) {
+                minIndex = currentIndex;
+                minRemainingValue = remaining;
+            }
+        }
+    
+        if (rateLimits[minIndex]) {
```
Comment (dgaubert) [inline lib/carto_metadata.js]: what about a ternary?
```
return rateLimits[minIndex] ? rateLimits[minIndex] : undefined;
```

Comment (oleurud) [reply lib/carto_metadata.js]: Showing your comment, I can do:
```
return rateLimits[minIndex]
```

Comment (dgaubert) [reply lib/carto_metadata.js]: do it! ;)

**Pull request merged**